### PR TITLE
[Refactor] Introduce connector sink operator for connector sink

### PR DESCRIPTION
**Pull request by letian-jiang**
## Why I'm doing:

Introduce connector sink operator for connector sink. Part of https://github.com/StarRocks/starrocks/pull/40542

## What I'm doing:

Implement a common sink operator used for all kinds of connector sink.

`ConnectorSinkOperator` manages two FIFO queue of `std::future`(s), which come from `ConnectorChunkSink`.

```cpp
    mutable std::queue<std::future<Status>> _add_chunk_future_queue;
    mutable std::queue<std::future<formats::FileWriter::CommitResult>> _commit_file_future_queue;
```

1. The operator cannot accept more chunks if any `add_chunk_future` is not ready. But it can still accept chunks if `commit_file_future` is not ready.
2. The operator is not finished if any `add_chunk_future` and `commit_file_future` is not ready.
3. The operator will invoke callback_function defined by `ConnectorChunkSink` once a `commit_file_future` is ready.

## What type of PR is this:

- [ ] BugFix
- [ ] Feature
- [ ] Enhancement
- [x] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [ ] 3.2
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

File: be/src/connector_sink/file_chunk_sink.cpp [hunk 1]
```diff
@@ -93,11 +94,11 @@ std::function<void(const formats::FileWriter::CommitResult& result)> FileChunkSi
     };
 }
 
-StatusOr<std::unique_ptr<ConnectorChunkSink>> FileChunkSinkProvider::create_chunk_sink(
+std::unique_ptr<ConnectorChunkSink> FileChunkSinkProvider::create_chunk_sink(
         std::shared_ptr<ConnectorChunkSinkContext> context, int32_t driver_id) {
     auto ctx = std::dynamic_pointer_cast<FileChunkSinkContext>(context);
     auto runtime_state = ctx->fragment_context->runtime_state();
-    ASSIGN_OR_RETURN(auto fs, FileSystem::CreateUniqueFromString(ctx->path, FSOptions(&ctx->cloud_conf)));
+    auto fs = FileSystem::CreateUniqueFromString(ctx->path, FSOptions(&ctx->cloud_conf)).value();
```
Comment (satanson) [inline be/src/connector_sink/file_chunk_sink.cpp:101]: is it a blocking operation executed in pipeline executor thread? if so, the blocking operations must be changed to  non-blocking style.

File: be/src/exec/pipeline/sink/connector_sink_operator.cpp [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+ConnectorSinkOperatorFactory::ConnectorSinkOperatorFactory(
+        int32_t id, std::unique_ptr<connector::ConnectorChunkSinkProvider> data_sink_provider,
+        std::shared_ptr<connector::ConnectorChunkSinkContext> sink_context, FragmentContext* fragment_context)
+        : OperatorFactory(id, "connector sink operator", Operator::s_pseudo_plan_node_id_for_final_sink),
+          _data_sink_provider(std::move(data_sink_provider)),
+          _sink_context(sink_context),
+          _fragment_context(fragment_context) {}
+
+OperatorPtr ConnectorSinkOperatorFactory::create(int32_t degree_of_parallelism, int32_t driver_sequence) {
+    auto chunk_sink = _data_sink_provider->create_chunk_sink(_sink_context, driver_sequence);
```
Comment (satanson) [inline be/src/exec/pipeline/sink/connector_sink_operator.cpp:149]: create_chunk_sink is called in fe->be brpc service handler threads, we must guarantee that this function would not  hold the thread too much time.

File: be/src/exec/pipeline/sink/connector_sink_operator.cpp [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+    __builtin_unreachable();
+}
+
+Status ConnectorSinkOperator::push_chunk(RuntimeState* state, const ChunkPtr& chunk) {
+    ASSIGN_OR_RETURN(auto future, _connector_chunk_sink->add(chunk));
+    _enqueue_futures(std::move(future));
+    return Status::OK();
+}
+
+void ConnectorSinkOperator::_enqueue_futures(connector::ConnectorChunkSink::Futures future) {
```
Comment (satanson) [inline be/src/exec/pipeline/sink/connector_sink_operator.cpp:131]: param name: connector::ConnectorChunkSink::Futures future, suggest to use futures.
Futures' fields add_chuck_future and commit_file_future, also suggest to use plurals.

File: be/src/connector_sink/file_chunk_sink.cpp [hunk 4]
```diff
@@ -41,6 +41,7 @@ FileChunkSink::FileChunkSink(std::vector<std::string> partition_columns,
 
 Status FileChunkSink::init() {
```
Comment (satanson) [inline be/src/connector_sink/file_chunk_sink.cpp:42]: this FileChunkSink is used by Compute-Separated-From-Storage?

File: be/src/connector_sink/hive_chunk_sink.cpp [hunk 5]
```diff
@@ -41,6 +42,7 @@ HiveChunkSink::HiveChunkSink(std::vector<std::string> partition_columns,
 
 Status HiveChunkSink::init() {
     RETURN_IF_ERROR(ColumnEvaluator::init(_partition_column_evaluators));
+    RETURN_IF_ERROR(_file_writer_factory->init());
```
Comment (satanson) [inline be/src/connector_sink/hive_chunk_sink.cpp:45]: FileWriter::commit and write seems blocking operations, it must be non-blocking.

Comment (letian-jiang) [reply be/src/connector_sink/hive_chunk_sink.cpp:45]: non-blocking

File: be/src/connector_sink/file_chunk_sink.cpp (hunk 1 above)
Comment (letian-jiang) [reply be/src/connector_sink/file_chunk_sink.cpp:101]: non-blocking. `CreateUniqueFromString` only invokes ctor of relevant file system.

File: be/src/exec/pipeline/sink/connector_sink_operator.cpp (hunk 3 above)
Comment (letian-jiang) [reply be/src/exec/pipeline/sink/connector_sink_operator.cpp]: renamed.

**Review by mxdzs0612 — APPROVED** [review]
LGTM

**Review by satanson — DISMISSED** [review]

File: be/src/exec/pipeline/sink/connector_sink_operator.cpp [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+    }
+
+    while (!_add_chunk_future_queue.empty()) {
+        // cannot accept chunk if any add_chunk_futures is not ready
+        if (!is_ready(_add_chunk_future_queue.front())) {
+            return false;
+        }
+        if (auto st = _add_chunk_future_queue.front().get(); !st.ok()) {
+            LOG(WARNING) << "cancel fragment: " << st;
+            _fragment_context->cancel(st);
```
Comment (imay) [inline be/src/exec/pipeline/sink/connector_sink_operator.cpp:62]: Will this operation be a time-cost operation? or block operation?

File: be/src/exec/pipeline/sink/connector_sink_operator.cpp [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+    return !is_finished();
+}
+
+Status ConnectorSinkOperator::set_cancelled(RuntimeState* state) {
+    _is_cancelled = true;
+    return Status::OK();
+}
+
+StatusOr<ChunkPtr> ConnectorSinkOperator::pull_chunk(RuntimeState* state) {
+    CHECK(false) << "ConnectorSinkOperator::pull_chunk";
```
[the remark this answers is not in the document]
Comment (letian-jiang) [reply be/src/exec/pipeline/sink/connector_sink_operator.cpp]: done

File: be/src/exec/pipeline/sink/connector_sink_operator.cpp (hunk 6 above)
Comment (letian-jiang) [reply be/src/exec/pipeline/sink/connector_sink_operator.cpp:62]: No. It is neither time-cost nor blocking.

File: be/src/exec/pipeline/sink/connector_sink_operator.cpp [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+        return false;
+    }
+
+    while (!_add_chunk_future_queue.empty()) {
+        // cannot accept chunk if any add_chunk_futures is not ready
+        if (!is_ready(_add_chunk_future_queue.front())) {
+            return false;
+        }
+        if (auto st = _add_chunk_future_queue.front().get(); !st.ok()) {
+            LOG(WARNING) << "cancel fragment: " << st;
```
[the remark this answers is not in the document]
Comment (letian-jiang) [reply be/src/exec/pipeline/sink/connector_sink_operator.cpp:61]: This is used to show what error causes the fragment to be canceled.

And I found similar usage in other operators (e.g. olap_table_sink_operator)
```cpp
    if (!_sink->is_close_done()) {
        auto st = _sink->try_close(_fragment_ctx->runtime_state());
        if (!st.ok()) {
            _fragment_ctx->cancel(st);
            return false;
        }
        return true;
    }
```

**Review by imay — APPROVED** [review]

**Review by satanson — APPROVED** [review]

**Pull request merged**